zfit/core/constraint.py

```diff
@@ -0,0 +1,183 @@
+#  Copyright (c) 2019 zfit
+
+import abc
+from collections import OrderedDict
+
+from typing import Dict, Union, Callable, Optional
+
+from .baseobject import BaseNumeric
+from .interfaces import ZfitParameter
+from ..util import ztyping
+from ..util.graph import get_dependents_auto
+from ..util.container import convert_to_container
+from ..util.exception import ShapeIncompatibleError
+from ..settings import ztypes
+from zfit import ztf
+import zfit
+
+import tensorflow as tf
+import numpy as np
+import tensorflow_probability as tfp
+
+tfd = tfp.distributions
+
+
+class BaseConstraint(BaseNumeric):
+
+    def __init__(self, params: Union[Dict[str, ZfitParameter]] = None,
+                 name: str = "BaseConstraint", dtype=ztypes.float,
+                 **kwargs):
+        """Base class for constraints.
+
+        Args:
+            dtype (DType): the dtype of the constraint
+            name (str): the name of the constraint
+            params (Dict(str, :py:class:`~zfit.Parameter`)): A dictionary with the internal name of the
+                parameter and the parameters itself the constrains depends on
+        """
+        super().__init__(name=name, dtype=dtype, params=params, **kwargs)
+
+    def value(self):
+        return self._value()
+
+    @abc.abstractmethod
+    def _value(self):
+        raise NotImplementedError
+
+    def _get_dependents(self) -> ztyping.DependentsType:
+        return self._extract_dependents(self.get_params())
+
+    def sample(self, n):
```

Comment:
no default for n, because n=1 is rather a special case than the vast majority of the cases (right?). So rather require it

Reply:
Ok 👍🏻